### PR TITLE
fix(router): don't break history when CanDeactivate cancel back navigation

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Closes #https://github.com/angular/angular/issues/13586

Description:
`location.back()` is called before router so there's no way to prevent if when `CanDeactivate` return `false`. 
To maintain proper history we need to push (`location.pushState`) new entry into the history instead of replacing (`location.replaceState`) the current entry.

History before the fix

| Action | Length | Current index | Url     |
|--------|--------|---------------|---------|
|  Init navigation      | 1      | 1             | /       |
|   Navigate /route1     | 2      | 2             | /route1 |
|  Navigate /route2      | 3      | 3             | /route2 |
|    Navigate /route3    | 4      | 4             | /route3 |
| Navigate back (CanDeactivate false)   | 4      | 3 (here we are replacing current state instead of pushing new entry, this is why index is 3 instead of 4)            | /route3 |
| Navigate back (CanDeactivate true) | 4      | 2             | /route1 |

After the fix

| Action | Length | Current index | Url     |
|--------|--------|---------------|---------|
|  Init navigation      | 1      | 1             | /       |
|   Navigate /route1     | 2      | 2             | /route1 |
|  Navigate /route2      | 3      | 3             | /route2 |
|    Navigate /route3    | 4      | 4             | /route3 |
| Navigate back (CanDeactivate false)   | 4      | 4             | /route3 |
| Navigate back (CanDeactivate true) | 4      | 3             | /route2 |

TLTR our goal: we need to call `location.pushState` instead of `location.replaceState` when `CanDeactivate` cancels `back` navigation

http://plnkr.co/edit/GMkEl74JkQiRPMv6Km5Z?p=preview

### CODE REVIEW

**Comment by mhevery** [comment]
@vicb Review when you get a chance.

**Comment by michelherv** [comment]
Hi,
I have the same problem in case of Activation Guard. Have I to create a new issue ?

* **Angular version:** 2.4.6
* **Browser:** all
* **Language:** TypeScript 2.1.5

**Comment by vicb** [comment]
@DzmitryShylovich could toy please rebase the PR ?
Thanks

File: modules/@angular/router/src/router.ts [hunk 1]
```diff
@@ -613,14 +613,6 @@ export class Router {
       Promise<boolean> {
     const lastNavigation = this.navigations.value;
 
-    // If the user triggers a navigation imperatively (e.g., by using navigateByUrl),
```
Comment (vicb) [inline modules/@angular/router/src/router.ts]: why is this removed ?

I don't think it relates to the issue ?

File: modules/@angular/router/src/router.ts [hunk 2]
```diff
@@ -796,8 +789,22 @@ export class Router {
                       id, this.serializeUrl(url), this.serializeUrl(this.currentUrlTree)));
                   resolvePromise(true);
                 } else {
-                  this.resetUrlToCurrentUrlTree();
-                  this.routerEvents.next(new NavigationCancel(id, this.serializeUrl(url), ''));
+                  const navigatedUrl: string = this.serializeUrl(url);
+                  // When navigation is canceled and it was initiated by the click on Back button,
```
Comment (vicb) [inline modules/@angular/router/src/router.ts]: What if was the forward button ?

**Review by vicb — CHANGES_REQUESTED** [review]
see inline comments

File: modules/@angular/router/src/router.ts (hunk 1 above)
Comment (DzmitryShylovich) [inline modules/@angular/router/src/router.ts]: this check breaks [this](https://github.com/angular/angular/pull/13922/files#diff-f48b128ef12ddeb072207e5ba88d04eeR1954) test.

Comment (DzmitryShylovich) [inline modules/@angular/router/src/router.ts]: there're no unit tests for this so I'm not sure what uses cases it covers

**Comment by DzmitryShylovich** [comment]
@vicb rebased
> What if was the forward button ?

if `forward` button was called but navigation was canceled by a guard browser history would be 1 item ahead. afaik hostory api doesn't allow to delete entries so to maintain a proper history we will need to call `location.back()`.
It's simple:
1. if back button was canceled we need to push 1 item by calling `location.pushState` because we're 1 item above
2. if forward button was canceled we need to remove 1 item by calling `location.back` because we're 1 item ahead

**Comment by DzmitryShylovich** [comment]
@vicb I've created a new issue for Location.forward bug https://github.com/angular/angular/issues/15664

**Review by bakhshi — APPROVED** [review]

**Comment by jasonaden** [comment]
Fixed moved to #18135

**Comment by blueberrynotblue** [comment]
feel sad for problem remained on 4.3

**Comment by UlricW** [comment]
This issue is it fixed ?
Thanks

**Comment by zvonimirprofico** [comment]
Does not seem like it is https://github.com/angular/angular/issues/13586#issuecomment-485036336

**Pull request closed without merging**